### PR TITLE
SwiftFormat

### LINKED ISSUE #887: swiftformat?
As long as I have some PRs hanging out there, here's another topic for discussion. Would you consider adopting [swiftformat](https://github.com/nicklockwood/SwiftFormat)? We use this quite happily in my current project alongside similar tools in (prettier, rubocop, etc).

Swiftformat would make it easier for newcomers (like me) to make contributions and not have to worry about whitespace/formatting issues. I spent a fair bit of time ironing out formatting issues for those PRs, unfortunately. I'm not familiar with Danger, but I imagine you could enforce via the Dangerfile and the PR checklist.

### PR DESCRIPTION
**Pull request by gurgeous**
Adds [SwiftFormat](https://github.com/nicklockwood/SwiftFormat), fixes #887. Here are some specific things to look at in the PR:

1. Initial set of .swiftformat rules. The presence of SwiftFormat is more important than the actual rules IMO.
1. Changes to .swiftlint.yml to ignore transient warnings that are fixed by running SwiftFormat.
1. New file `scripts/pre-commit`, a pre-commit hook that runs SwiftFormat. It includes installation instructions.
1. Modifications to swiftlint:disable statements to work around wrapping issues.

Future work to consider:

1. Document swiftformat for core devs.
1. Add a Brewfile. Swiftlint should go in there too.
1. Perhaps we should add `tx, ty, tz, sx, sy, sz, & t2` to the identifier_name exclusions.

### CODE REVIEW

**Comment by gurgeous** [comment]
CI was complaining about the `.swiftversion` file, so I switched to `--swiftversion 5.1` in `.swiftformat`. This matches the podspec. Moving swiftformat to 5.1 also enabled it to ditch all those unnecessary return keywords.

File: Sources/SwifterSwift/AppKit/NSImageExtensions.swift
Comment (LucianoPAlmeida) [inline Sources/SwifterSwift/AppKit/NSImageExtensions.swift:55]: humm... this formating style felt really weird to me, I would expect something like breaking at the begining

File: Sources/SwifterSwift/AppKit/NSViewExtensions.swift
Comment (LucianoPAlmeida) [inline Sources/SwifterSwift/AppKit/NSViewExtensions.swift:32]: Personally, I'm not a big fan of this enforcement ... I think although the language allows this, I think we should use in very specific case(personally I would use it more in cases where we could do one-liner e.g. var x { a + b } ). But I'll let you guys decide on that :) To me if we would enforce this I think we should enforce at least to try let the single expression in the same line e.g. `get { layer?.borderWidth ?? 0 }` I think it would be a good way to go

File: Sources/SwifterSwift/Dispatch/DispatchQueueExtensions.swift
Comment (LucianoPAlmeida) [inline Sources/SwifterSwift/Dispatch/DispatchQueueExtensions.swift:49]: Is this a bug, maybe line width maybe conflicting with the same-line bracet rule ... ?

File: Sources/SwifterSwift/Foundation/NSRegularExpressionExtensions.swift
Comment (LucianoPAlmeida) [inline Sources/SwifterSwift/Foundation/NSRegularExpressionExtensions.swift:25]: I like this very much! Line width is nice

File: Sources/SwifterSwift/AppKit/NSImageExtensions.swift
Comment (gurgeous) [reply Sources/SwifterSwift/AppKit/NSImageExtensions.swift:55]: See: https://github.com/nicklockwood/SwiftFormat/blob/master/Rules.md#elseonsameline

I personally like `--guardelse same-line`, though I also don't mind sticking to defaults.

image

File: Sources/SwifterSwift/AppKit/NSViewExtensions.swift
Comment (gurgeous) [reply Sources/SwifterSwift/AppKit/NSViewExtensions.swift:32]: See: https://github.com/nicklockwood/SwiftFormat/blob/master/Rules.md#redundantreturn

I think this is idiomatic Swift now and thus the default with SwiftFormat. Once the redundant returns are gone the remaining returns are clear indicators of early exit. I personally prefer this style, especially after working in other languages with implicit return (like Ruby).

Not a strong opinion, though. We can disable the `redundantReturn` rule if others agree. @guykogus? @omaralbeik? Chime in if you have opinions...

image

File: Sources/SwifterSwift/Dispatch/DispatchQueueExtensions.swift
Comment (gurgeous) [reply Sources/SwifterSwift/Dispatch/DispatchQueueExtensions.swift:49]: I suspect it always uses a newline when parameter are wrapped. Probably to give a cleaner indent after the bracket. Not sure, though.

File: Sources/SwifterSwift/AppKit/NSViewExtensions.swift
Comment (LucianoPAlmeida) [reply Sources/SwifterSwift/AppKit/NSViewExtensions.swift:32]: Yeah, I like it very much but in a case by case basis, as I mentioned we should have a convention to use as we see fit, but not enforce as a rule everywhere :)

**Comment by gurgeous** [comment]
> Here are my 2 cents overall
> 
> 1. In general looks nice, I just have a minor concern about enforcing omitting the return. See comments bellow for details but TBH I would prefer not to enforce this.

This is worth a bit of discussion IMO. Let's give it a day and see if anyone else chimes in, otherwise we can disable it. It would probably be some work to restore the missing returns since running SwiftFormat again won't magically bring them back in the PR :)


> 2. There are some places where maxwidth restrictions force it to put the bracket in the line bellow which conflicts with `closingparen same-line` is that a bug?

I don't think so - see above. Let me know if you think this merits further investigation.


> 3. I'm not a big fan of those big refactors because I think this messes up with git history ... at least makes it a bit harder to git blame and track down PRs/commits where things were really implemented. I struggled a bit with this in the past. But I think there is no way around this right? any alternatives?

The perfectionist in me finds them frustrating as well, but the pragmatist just pushes on through. :) Despite my misgivings it's usually not much of a problem for projects like this. Especially when all the methods are nicely self contained and git blame is rarely required.

 
> But overall it looks great! Thank you for taking this on @gurgeous!

Thank you, my pleasure!! This is a fun and important project and I look forward to doing more.

**Comment by gurgeous** [comment]
No further comments on this one, so let's finish it up. @LucianoPAlmeida let me know if you want me to disable `redundantReturn`. That will require redoing the PR (to restore the missing `returns`) but I'd be happy to tackle it if you think worthwhile.

Should we add a line to the changelog? It's mostly internal, not sure anyone will care apart from core devs.

**Comment by LucianoPAlmeida** [comment]
> No further comments on this one, so let's finish it up. @LucianoPAlmeida let me know if you want me to disable `redundantReturn`. That will require redoing the PR (to restore the missing `returns`) but I'd be happy to tackle it if you think worthwhile.
> 

I was actually waiting on thoughts from the others @omaralbeik @guykogus, I'm +1 to disable it, but we normally like to have more opinions on stuff like this :)

> Should we add a line to the changelog? It's mostly internal, not sure anyone will care apart from core devs.

IMH no need for changelog here since doesn't affect users.

File: Sources/SwifterSwift/Foundation/NSRegularExpressionExtensions.swift
Comment (guykogus) [inline Sources/SwifterSwift/Foundation/NSRegularExpressionExtensions.swift:89]: This is what I meant before about redundant returns for functions. I find it very strange.

File: .swiftformat
Comment (guykogus) [inline .swiftformat:8]: Aren't we on 5.2?

**Review by guykogus — COMMENTED** [review]
Good stuff! I don't love all the formatting but it's great to have something.
Also, is there some way to set up GitHub Actions to run swiftformat and ensure that there are no changes when calling `git status`? I did something like that at the last company I worked for and my old colleagues still hate me for it 😅

File: Sources/SwifterSwift/AppKit/NSImageExtensions.swift
Comment (gurgeous) [reply Sources/SwifterSwift/AppKit/NSImageExtensions.swift:55]: Let's try `--guardelse same-line`

File: Sources/SwifterSwift/AppKit/NSViewExtensions.swift
Comment (gurgeous) [reply Sources/SwifterSwift/AppKit/NSViewExtensions.swift:32]: I don't think we can customize it to that extent. Personally this doesn't bother me (because Ruby) but I don't have a strong opinion. Our convention is good even without autoformatting, I think. Let's turn off `redundantReturn`

File: Sources/SwifterSwift/Dispatch/DispatchQueueExtensions.swift
Comment (gurgeous) [reply Sources/SwifterSwift/Dispatch/DispatchQueueExtensions.swift:49]: Maybe we can disable `wrapMultilineStatementBraces`

File: Sources/SwifterSwift/CoreLocation/CLLocationArrayExtensions.swift
Comment (gurgeous) [reply Sources/SwifterSwift/CoreLocation/CLLocationArrayExtensions.swift:19]: Might be able to customize this with judicious use of `--nospaceoperators`. I will experiment and if it looks promising add to config.

File: Sources/SwifterSwift/Foundation/DateExtensions.swift
Comment (gurgeous) [reply Sources/SwifterSwift/Foundation/DateExtensions.swift:54]: 👍

File: Sources/SwifterSwift/CoreAnimation/CATransform3DExtensions.swift
Comment (gurgeous) [reply Sources/SwifterSwift/CoreAnimation/CATransform3DExtensions.swift:163]: Here are a few ideas, not in order:

1. Move the comment up
1. Add sx/sy/sz/tx/ty/tz/t2... to our list of exclusions in the config.
1. Disable rule for the entire file.

Pick one!

File: .swiftformat
Comment (gurgeous) [reply .swiftformat:8]: The podspec said 5.1, but I think we are building 5.2. I'll change it.

**Comment by gurgeous** [comment]
The branch has been updated with swiftformat config changes in response to the PR feedback. I used the following:

```sh
$ git checkout upstream/master -- Sources Tests
$ swiftformat .
```

There are additional fixes that need to happen once we're happy with the formatting:

1. fix `swiftlint:disable identifier_name` after we pick a fix
1. the ugly wrapping in `DateExtensions`

I'll tackle those once we are done with the formatting. @LucianoPAlmeida & @guykogus can you take another look? I think this is getting close. Thanks!

**Review by guykogus — CHANGES_REQUESTED** [review]
For CATransform3DExtensions let's just disable `identifier_name` so that we don't have to mark it for each instance.

File: Sources/SwifterSwift/CoreAnimation/CATransform3DExtensions.swift
Comment (guykogus) [reply Sources/SwifterSwift/CoreAnimation/CATransform3DExtensions.swift]: 3

**Comment by gurgeous** [comment]
If there are no further comments I will apply those last two changes and prep for merge. @guykogus not sure about CI but I'd suggest we save that for a subsequent effort. I haven't used Github Actions yet... Hopefully we can start with the pre-commit hook. :) I also think we should add a Brewfile, but that can wait as well.

File: Brewfile
Comment (guykogus) [inline Brewfile:3]: No need for this comment
```suggestion
brew "swiftformat"
```

File: scripts/pre-commit
Comment (guykogus) [inline scripts/pre-commit:10]: I guess for us it's from the home folder, no?
```suggestion
# $ ln -sf scripts/pre-commit .git/hooks/pre-commit
```

**Review by guykogus — CHANGES_REQUESTED** [review]
Looking great. Just a couple of little comments

File: scripts/pre-commit
Comment (gurgeous) [reply scripts/pre-commit:10]: I think it has to be relative to the location of `pre-commit`:

```sh
$ cd SwifterSwift
$ ln -sf scripts/pre-commit .git/hooks/pre-commit
$ .git/hooks/pre-commit
command not found: .git/hooks/pre-commit

$ cd SwifterSwift
$ ln -sf ../../scripts/pre-commit .git/hooks/pre-commit
$ .git/hooks/pre-commit
... runs properly ...
```

I find this counterintuitive as well and I messed it up the first time we tried this with our app 😄

Comment (guykogus) [reply scripts/pre-commit:10]: Ah I see, in the first instance you're creating a link that's trying to find `.git/hooks/scripts/pre-commit`, instead of `.git/hooks/../../scripts/pre-commit`. Makes sense now.

**Review by guykogus — APPROVED** [review]

**Comment by LucianoPAlmeida** [comment]
cc @omaralbeik

**Review by omaralbeik — APPROVED** [review]
Nice work 🚀

**Pull request merged**
